jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,149 @@
+    // meeting remain
+    List<TimeRange> validTimeRanges = getValidTimeRanges(openTimeRanges, (int) request.getDuration());
+
+    return validTimeRanges;
   }
+
+  // Adds all relevant events (events with at least one required attendee) to a new collection.
+  private Collection<Event> getRelevantEvents(Collection<Event> events, Collection<String> requiredAttendees) {
+    Collection<Event> relevantEvents = new HashSet<>();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+
+      if (!Collections.disjoint(requiredAttendees, eventAttendees)) {
+        relevantEvents.add(event);
+      }
+    }
+
+    return relevantEvents;
+  }
+
+  // Adds the timeranges of all events in the collection to a list, sorted by start time.
+  private List<TimeRange> getEventTimeRanges(Collection<Event> events) {
+    List<TimeRange> timeRanges = new ArrayList<>();
+
+    for (Event event : events) {
+      timeRanges.add(event.getWhen());
+    }
+
+    // Sort the timeRanges by start time.
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    return timeRanges;
+  }
+
+  /**
+   * Condenses an ordered list of time ranges (ordered by start time).
+   * @param timeRanges list of timeRanges sorted by start time
+   * @return list of condensed timeRanges (condensed means overlapping events are turned into a single event).
+   */
+  private List<TimeRange> getCondensedTimeRanges(List<TimeRange> timeRanges) {
+    List<TimeRange> condensedTimeRanges = new ArrayList<>();
+
+    for (TimeRange timeRange : timeRanges) {
+      // Case 0: no previous timeRanges
+      if (condensedTimeRanges.isEmpty()) {
+        condensedTimeRanges.add(timeRange);
+
+      } else {
+        TimeRange currentCondensedRange = condensedTimeRanges.get(condensedTimeRanges.size()-1);
+
+        // Case 1: current timeRange does not overlap with current condensed time range being built
+        if (!currentCondensedRange.overlaps(timeRange)) {
+          // Add the timeRange as a new condensed range
+          condensedTimeRanges.add(timeRange);
+        }
+        // Case 2: current timeRange overlaps with current condensed time range being built, but is not contained in it
+        else if (currentCondensedRange.overlaps(timeRange) && !currentCondensedRange.contains(timeRange)) {
+          // Replace the current condensed time range with a new condensed time range augmented with the current timeRange
+          TimeRange newCondensedRange = TimeRange.fromStartEnd(currentCondensedRange.start(), timeRange.end(), false);
+
+          condensedTimeRanges.remove(condensedTimeRanges.size()-1);
+          condensedTimeRanges.add(newCondensedRange);
+        }
+
+        // Case 3: if current timeRange is contained within current condensed time range being built, do nothing
+      }
+    }
+
+    return condensedTimeRanges;
+  }
+
+  /**
+   * Finds the timeRanges in the day which are not covered by any events. May add ranges of duration 0
```

Comment:
Perhaps you could clarify `May add ranges of duration 0 between timeRanges.`

Do you just mean that some ranges may have a duration of 0?

Reply:
I clarified this point: an inverse range of duration zero will be added to represent the space between adjacent events.